### PR TITLE
Remove awesome_nested_set gem

### LINKED ISSUE #11636: Remove gem awesome_nested_set
### What we should change and why (this is tech debt)

It's providing a tree structure for Spree::Taxon but we don't use the tree structure. I think that the UI is actually not saving it when you try. If we ever wanted a tree structure in the future, it would be better to use [ancestry](https://github.com/stefankroes/ancestry) for it's simpler design and better update performance.

There are several related issues that may be resolved at the same time:

* #9677
* #9386
* #6033

### Context

* #11631

### Impact and timeline

No significant impact at the moment. But simplifying code always reduces maintenance and performance.

### PR DESCRIPTION
**Pull request by HillaryOkello**
- Closes #11636

#### Release notes

Changelog Category (reviewers may add a label for the release notes):

- [x] Technical changes only

The title of the pull request will be included in the release notes.

### CODE REVIEW

File: app/models/spree/taxon.rb [hunk 1]
```diff
@@ -21,19 +21,15 @@ def applicable_filters
...
-      if parent.present?
-        self.permalink = [parent.permalink, permalink_end].join('/')
-      elsif permalink.blank?
-        self.permalink = UrlGenerator.to_url(name)
-      end
+      self.permalink = if parent.present?
+                         [parent.permalink, permalink_end].join('/')
+                       elsif permalink.blank?
+                         UrlGenerator.to_url(name)
+                       end
```
Comment (mkllnk) [inline app/models/spree/taxon.rb:33]: There are a few formatting changes in this commit which seem unrelated. It's better to commit these separately to show clearly what the intention of the commit is.

File: app/models/spree/taxon.rb [hunk 2]
```diff
@@ -42,10 +38,8 @@ def to_param
     end
 
     def pretty_name
-      ancestor_chain = ancestors.inject("") do |name, ancestor|
-        name + "#{ancestor.name} -> "
-      end
-      ancestor_chain + name.to_s
+      ancestor_chain = ancestors.map(&:name).join(' -> ')
```
Comment (mkllnk) [inline app/models/spree/taxon.rb:42]: It looks like `ancestors` was a method provided by the removed gem. So it can't be used here. Maybe it's enough to just refer to the first parent?

File: app/models/spree/taxon.rb [hunk 3]
```diff
@@ -91,9 +85,7 @@ def self.distributed_taxons(which_taxons = :all)
     private
 
     def permalink_end
-      return UrlGenerator.to_url(name) if permalink.blank?
-
-      permalink.split('/').last
+      permalink.blank? ? UrlGenerator.to_url(name) : permalink.split('/').last
```
Comment (mkllnk) [inline app/models/spree/taxon.rb:89]: Why did you change this? Ternary operators are hard to read.

**Review by mkllnk — CHANGES_REQUESTED** [review]
Thank you for your contribution. Amongst so many failing specs it's difficult to find the root cause. A serializer spec is telling me that the `ancestors` method doesn't exist. Try fixing that and then see how many specs that fixes.

File: app/models/spree/taxon.rb (hunk 1 above)
Comment (HillaryOkello) [reply app/models/spree/taxon.rb:33]: I'll revert this

File: app/models/spree/taxon.rb (hunk 3 above)
Comment (HillaryOkello) [reply app/models/spree/taxon.rb:89]: I'll revert this as well

**Review by mkllnk — COMMENTED** [review]
Nice work. It looks like you got everything working again.

The issue said:

> we don't use the tree structure.

So I was hoping that we can get rid of a lot of functionality and don't need to add code to remove the gem. Since you are new to the project, you don't know how this code is used and how much you can remove. Some of the specs may need adjusting as well.

While you've made very good progress, I'm not sure if you are happy to continue. You would need to:

* Search through the code to find where taxons are used.
* Find out if it's important that the taxons still have an order (like a list). If so, we need to replace the list functionality. Otherwise we can get rid of it.
* Replace all code using the tree structure (like ancestors) and assume a flat hierarchy. In production, we don't have a tree here.

In the end, we want less code than before. Do you think that you can do that?

**Comment by HillaryOkello** [comment]
@mkllnk I'm happy to continue. I'll work on making the suggestions you requested. In case I get stuck I'll ask for assistance

**Comment by mkllnk** [comment]
> I'm happy to continue.

Great, I'm looking forward to seeing your progress.

Just as a side note, we prefer to rebase on the master branch instead of merging while a pull request is still in development. So feel free to rebase and maybe squash some commits to follow our guide on good pull requests:

* https://github.com/openfoodfoundation/openfoodnetwork/wiki/Making-a-great-commit
* https://github.com/openfoodfoundation/openfoodnetwork/wiki/Making-a-great-pull-request
